### PR TITLE
add Index::builder, add index_settings to IndexMeta

### LINKED ISSUE #1018: Add builder API to create Index/IndexMeta
As prerequisite for #1014, we need an API layer above `Schema` to set global options like index sorting

### PR DESCRIPTION
**Pull request by PSeitz**
fixes #1018

### CODE REVIEW

File: src/indexer/segment_updater.rs [hunk 1]
```diff
@@ -46,6 +47,7 @@ const NUM_MERGE_THREADS: usize = 4;
 pub fn save_new_metas(schema: Schema, directory: &dyn Directory) -> crate::Result<()> {
     save_metas(
         &IndexMeta {
+            index_settings: IndexSettings::default(),
```
Comment (PSeitz) [inline src/indexer/segment_updater.rs]: @fulmicoton 
Can we load the index metas from Index to access index_settings or does it not yet exist?

If it doesn't exist, we would need to either write them upfront or move IndexMeta into Index to have a place to store index settings on creation.

File: src/core/index.rs [hunk 2]
```diff
@@ -55,6 +55,66 @@ fn load_metas(
...
+/// let body_field = schema_builder.add_text_field("body", TEXT);
+/// let schema = schema_builder.build();
+/// let settings = IndexSettings::default();
+/// let index = Index::builder().schema(schema).settings(settings).create_in_ram();
+///
+/// ```
+pub struct IndexBuilder {
+    index_meta: IndexMeta,
+}
+impl Default for IndexBuilder {
```
Comment (fulmicoton) [inline src/core/index.rs]: do we use default?

File: src/indexer/segment_updater.rs (hunk 1 above)
Comment (fulmicoton) [reply src/indexer/segment_updater.rs]: I would give it the same status as schema.

pass it as an argument in `save_new_metas` and load it from IndexMeta and store it as a member of Index.

File: src/indexer/segment_updater.rs [hunk 3]
```diff
@@ -204,6 +206,7 @@ pub fn merge_segments<Dir: Directory>(
     );
 
     let index_meta = IndexMeta {
+        index_settings: indices[0].load_metas()?.index_settings, // index_settings of all segments should be the same
```
Comment (fulmicoton) [inline src/indexer/segment_updater.rs:217]: let's hold a IndexSettings member in Index (like we do with schema)

File: src/core/index.rs (hunk 2 above)
Comment (PSeitz) [reply src/core/index.rs]: removed

File: src/core/index.rs [hunk 4]
```diff
@@ -55,17 +55,142 @@ fn load_metas(
...
+        self.schema = Some(schema);
+        self
+    }
+    /// Creates a new index using the `RAMDirectory`.
+    ///
+    /// The index will be allocated in anonymous memory.
+    /// This should only be used for unit tests.
+    pub fn create_in_ram(self) -> Result<Index, TantivyError> {
+        let ram_directory = RAMDirectory::create();
+        Ok(Index::create(
```
Comment (fulmicoton) [inline src/core/index.rs:105]: So we you ended up not inversing the logic?

File: src/core/index.rs [hunk 5]
```diff
@@ -147,36 +259,34 @@ impl Index {
...
-        let mmap_directory = MmapDirectory::create_from_tempdir()?;
-        Index::create(mmap_directory, schema)
+        IndexBuilder::new().schema(schema).create_from_tempdir()
     }
 
     /// Creates a new index given an implementation of the trait `Directory`.
     ///
     /// If a directory previously existed, it will be erased.
-    pub fn create<Dir: Directory>(dir: Dir, schema: Schema) -> crate::Result<Index> {
+    pub fn create<Dir: Directory>(
```
Comment (fulmicoton) [inline src/core/index.rs:265]: can we move this guy as a private function of the Builder, and keep the original create for backward compatibility.

File: src/core/index.rs [hunk 6]
```diff
@@ -147,36 +259,34 @@ impl Index {
...
+        save_new_metas(schema.clone(), settings.clone(), &directory)?;
+        let mut metas = IndexMeta::with_schema(schema);
+        metas.index_settings = settings;
+        let index = Index::open_from_metas(directory, &metas, SegmentMetaInventory::default());
         Ok(index)
     }
 
     /// Creates a new index given a directory and an `IndexMeta`.
-    fn create_from_metas(
+    fn open_from_metas(
```
Comment (fulmicoton) [inline src/core/index.rs:278]: yes that's a more sensible name.

**Comment by fulmicoton** [comment]
LGTM apart for minor comment

File: src/core/index.rs (hunk 5 above)
Comment (PSeitz) [reply src/core/index.rs:265]: yes that's better. done

Comment (fulmicoton) [reply src/core/index.rs:265]: I do not see the change. Is it fixed?

Comment (PSeitz) [reply src/core/index.rs:265]: Yes, the last commit: https://github.com/tantivy-search/tantivy/pull/1021/commits/b9b9e9e518178ca07ce79a2bbd0cde5ba45064ab

**Pull request merged**